src/advanced_unsafety/uninitialized.md

```diff
@@ -0,0 +1,128 @@
+# Uninitialized memory
+
+> _"I'm Nobody! Who are you? Are you — Nobody — too?"_
+>
+> — _Emily Dickinson_
+
+While we have covered [invalid values], there's another thing that behaves a lot like invalid values, but has nothing to do with actual bit patterns: Uninitialized memory.
+
+An easy way to think about uninitialized memory is that there's an additional value (often called `undef` using LLVM's term for it) that does not map to any concrete bit pattern, but can be introduced in abstract in various ways, and makes _most_ values invalid.
+
+If you explicitly wish to work with uninitialized and partially-initialized types, [`MaybeUninit<T>`] is a useful abstraction since it can be "initialized" with no overhead and then written to in parts.
+
+## Sources of uninitialized memory
+
+### `mem::uninitialized()` and `MaybeUninit::assume_init()`
+
+[`mem::uninitialized()`] is a deprecated API that has a very tempting shape, it lets you do things like `let x = mem::uninitialized()` for cases when you want to construct the value in bits. It's basically _always_ UB to use, since it immediately sets `x` to uninitialized memory, which is UB.
+
+Use [`MaybeUninit<T>`] instead.
+
+It is still possible to create uninitialized memory using [`MaybeUninit::assume_init()`] if you have not, in fact, assured that things are initialized.
+
+### Padding
+
+Padding bytes in structs and enums are often but not always uninitialized. This means that treating a struct as a bag of bytes (by, say, treating `&Struct` as `&[u8; size_of::<Struct>()]` and reading from there) is UB even if you don't write invalid values to those bytes, since you are ginning up uninitialized `u8`s.
+
+Reading from padding [always produces uninitialized values][pad-glossary].
+
+
+
+### Moved-from values
+
+The following code is UB:
+
+```rust
+# use std::ptr;
+let x = String::new(); // String is not Copy
+let mut v = vec![];
+let ptr = &x as *const String;
+
+v.push(x); // move x into the vector
+
+unsafe {
+    // reads from moved-from memory
+    let ghost = ptr::read(ptr);
+}
+```
+
+Any type of move will do this, even when you "move" the value into a different variable with stuff like `let y = x;`.
+
+Note that Rust does let you "partially move" out of fields of a struct, in such a case the whole struct is now no longer a valid value for its type, but you are still allowed to "use" the struct to look at other fields. When doing such things, make sure there are no pointers that still think the struct is whole and valid.
```

Comment:
Might be worth mentioning that performing a partial move from a value makes the value _as a whole_ inaccessible. For example:

```rust
let x = (Foo::new(), Foo::new());
// After this assignment, what we call `x.0` is not actually a field
// of `x` any more because `x` doesn't exist any more.
let y = x.0;
// We can access and move out of `x.1` here even though `x`
// is no longer a valid `(Foo, Foo)` because `x` doesn't exist.
// What we call `x.1` is really just some `Foo` that was part of `x`.
let z = x.1;
```

Which is why you can't pass `x` anywhere or take references to it after performing a partial move.